File: MergeSort.java
public class Solution {


	public void sortIntegers(int[] A) {
		if (A == null || A.length == 0) {
			return;
		}

		// 用一个数组去存放中间的结果
		// 一边merge一边把结果先写在这个数组里面
		// 最后再写回到A数组中
		int[] temp = new int[A.length];

		mergeSort(A, 0, A.length -1, temp);
	}

	private void mergeSort(int[] A, int start, int end, int[] temp) {

		if (start >= end) {
			return;
		}

		// 处理左半区间
		mergeSort(A, start, (start + end)/2, temp);
		// 处理右半区间
		mergeSort(A, (start + end)/2 + 1, end, temp);

		// 合并
		merge(A, start, end, temp);
	}

	private void merge(int[] A, int start, int end, int[] temp) {

		int middle = (start + end)/2;
		int left = start;
		int right = middle + 1;

		int index = 0;

		while (left <= middle && right <= end) {

			if (A[left] > A[right]) {
				temp[index] = A[right];
				index ++;
				right ++;
			} else {
				temp[index] = A[left];
				index ++;
				left ++;
			}
		}

		while (left <= middle) {
			temp[index] = A[left];
			index ++;
			left ++;
		}

		while (right <= end) {
			temp[index] = A[right];
			index++;
			right++;
		}

		for (int i = start; i <= end; i ++) {
			A[i] = temp[i];		}
	}
}

File: merge_sort.py
class Solution:
	def sort(self, A):
		if not A:
			return A
		self.merge_sort(A, 0, len(A) -1)


    # [2, 1]
def mergeSort(self, A, start, end):
  if start >= end:
    return
  
  mid = (start + end) // 2
  self.mergeSort(A, start, mid)
  self.mergeSort(A, mid + 1, end)
  self.merge(A, start, mid, end)

def merge(self, A, start, mid, end):
  sorted_A = []
  left, right = start, mid + 1
  while left <= mid and right <= end:
    if A[left] < A[right]:
      sorted_A.append(A[left])
      left += 1
    else:
      sorted_A.append(A[right])
      right += 1	
  while left <= mid:
    sorted_A.append(A[left])
    left += 1
  while right <= end:
    sorted_A.append(A[right])
    right += 1
  # copy to original array
  for i in range(start, end + 1):
    A[i] = sorted_A[i - start]




	
